language-plutus-core/test/Language/PlutusCore/PropTest.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE RankNTypes        #-}
+
+module Language.PlutusCore.PropTest
+  ( TyProp
+  , testTyProp
+  , toClosedType
+  , normalizeTypeG
+  , GenError (..)
+  , ErrorP (..)
+  ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Gen.Common
+import           Language.PlutusCore.Gen.Type   hiding (toClosedType)
+import qualified Language.PlutusCore.Gen.Type   as Gen
+import           Language.PlutusCore.Pretty
+
+import           Control.Monad.Except
+import           Control.Search
+import qualified Data.Coolean                   as Cool
+import qualified Data.Text                      as Text
+import           Test.Tasty.HUnit
+import           Text.Printf
+
+
+-- |The type for properties with access to both representations.
+type TyProp =  Kind ()                           -- ^ kind for generated type
+            -> ClosedTypeG                       -- ^ generated type
+            -> ExceptT GenError Quote (Type TyName DefaultUni ()) -- ^ external rep. of gen. type
+            -> Cool                              -- ^ whether the property holds
+-- |Internal version of type properties.
+type TyPropG =  Kind ()      -- ^ kind of the generated type
+             -> ClosedTypeG  -- ^ generated type
+             -> Cool         -- ^ whether the property holds
+
+
+-- |Test property on well-kinded types.
+testTyProp :: Int     -- ^ Search depth
+           -> Kind () -- ^ Kind for generated types
+           -> TyProp  -- ^ Property to test
+           -> IO ()
+testTyProp depth k typrop = do
+  -- NOTE: Any strategy which attempts fairness will crash the search!
+  --       These strategies evaluate !=> in *parallel*, and hence attempt
+  --       to convert ill-kinded types, but `toType` is partial!
+  -- TODO: This *may* be a bug in the lazy-search library.
+  -- UPDATE: toType is nolonger partial
+  result <- ctrex' O depth (toTyPropG typrop k)
+  case result of
+    Left  _   -> return ()
+    Right tyG -> assertFailure (errorMsgTyProp k tyG)
+
+
+-- |Generate the error message for a failed `TyProp`.
+errorMsgTyProp :: Kind () -> ClosedTypeG -> String
+errorMsgTyProp kG tyG =
+  case run (toClosedType kG tyG) of
+    Left (Gen ty k) ->
+      printf "Test generation error: convert type %s at kind %s" (show ty) (show k)
+    Right ty ->
+      case run (inferKind defConfig ty) of
+        Left err ->
+          printf "Counterexample found: %s, generated for kind %s\n%s"
+            (show (pretty ty)) (show (pretty kG)) (show (prettyPlcClassicDef (err :: TypeError DefaultUni ())))
+        Right k2 ->
+          printf "Counterexample found: %s, generated for kind %s, has inferred kind %s"
+            (show (pretty ty)) (show (pretty kG)) (show (pretty k2))
+  where
+    run :: ExceptT e Quote a -> Either e a
+    run = runQuote . runExceptT
+
+-- |Convert a `TyProp` to the internal representation of types.
+toTyPropG :: TyProp -> TyPropG
+toTyPropG typrop kG tyG = tyG_ok Cool.!=> typrop_ok
```

Comment:
So this is where we throw a test case away if the generated type is not well-kinded, right? In Zonk we record discarded tests like that:

```
      Rational:                                                            OK (8.78s)
        +++ OK, passed 5000 tests; 6618 discarded.
```

'cause it's necessary to know how many tests have actual tested something and how much time was spent on generating garbage. I believe we should have a similar kind of output here, because at the moment it's not clear what things are being generated (and it'd be nice to see some actual test case examples).

Reply:
See [my other comment](https://github.com/input-output-hk/plutus/pull/2077#discussion_r455402449). There are no discarded tests. The term generation is *exhaustive* up to the given size.

Reply:
That sounds extremely sensible.